\documentclass[smaller]{beamer}
\usepackage{colortbl}

\include{slide-macros}

% This file is a solution template for:

% - Giving a talk on some subject.
% - The talk is between 15min and 45min long.
% - Style is ornate.



% Copyright 2004 by Till Tantau <tantau@users.sourceforge.net>.
%
% In principle, this file can be redistributed and/or modified under
% the terms of the GNU Public License, version 2.
%
% However, this file is supposed to be a template to be modified
% for your own needs. For this reason, if you use this file as a
% template and not specifically distribute it as part of a another
% package/program, I grant the extra permission to freely copy and
% modify this file as you see fit and even to delete this copyright
% notice. 

\def\typezero{\circ}
\def\typeone{\bullet}

\setbeamertemplate{blocks}[default]
\usecolortheme{rose}

\mode<presentation>
{
  % \usetheme{default}
  \usetheme{boxes}
  % or ...
  \usefonttheme[options]{structuresmallcapsserif}

  \setbeamercovered{transparent}
  % or whatever (possibly just delete it)
}


\usepackage[english]{babel}
% or whatever

\usepackage[latin1]{inputenc}
% or whatever

\usepackage{times}
\usepackage[T1]{fontenc}
% Or whatever. Note that the encoding and the font should match. If T1
% does not look nice, try deleting the line with the fontenc.


\title[Context-dependent models] % (optional, use only with long paper titles)
{Context-dependent models of nucleotide substitution}

\author % (optional, use only with lots of authors)
{Peter Ralph}
% - Use the \inst{?} command only if the authors have different
%   affiliation.

\institute[USC]
{
  USC -- Computational Biology and Bioinformatics
  }
% - Use the \inst command only if there are several affiliations.
% - Keep it simple, no one is interested in your street address.

\date % (optional)
{October 14, 2013}


\begin{document}


% Abstract: 
% It is well-known that the nucleotide mutation process is context-dependent:
% the probability of change at a particular base depends not only on the
% nucleotide at that position, but also at its neighbors; CpG hypermutability
% is the most well-known such effect.  Selection for or against particular
% motifs adds more layers of context dependencies to the nucleotide
% substitution process.  Incorporating context dependence is difficult, because
% this in principle requires summing over all possible substitutions at all
% sites.  I will present a computaionally feasible method to do these
% calculations correctly, which enables Bayesian inference of phylogenetic
% branch lengths, mutation rates, and selection coefficients of short motifs.
% I will also illustrate with examples to some models of statistical physics,
% such as estimation of the magnetization and temperature in the Ising model.

%%%%
% OUTLINE
%
% Intro to the problem
%   statement of the problem
%      mutation + selection
%   artificial example: TASEP
%   propagation of dependency
%   e^{tG}
%   previous work
%
% Method
%   restrict to pattern frequencies: partial likelihood
%   still missing edge effects: marginalize over window
%     node average boundary
%   asymptotically correct: propagation of dependency
%   likelihood function
%   computation: sparse matrices
%   phylogenetic peeling (?)
%
% Applications
%   TASEP
%   CpG
%   Ising
%   real data?
%   ancestral state reconstruction


%%%%%%%%% INTRO
\section{Introduction \& Background}

\begin{frame}{The problem: context dependence}

  {\struct Data:} Long (nucleotide) sequences separated by (evolutionary) time \\
    with known (phylogenetic) relationship.

  {\struct Question:} Infer properties of the model:\\
    substitution rates; relative amounts of time.

  {\struct Example:} Each site evolves independently \ldots

\end{frame}


\begin{frame}{Substitution is mutation \& selection}

  formal description

}

% TASEP; propagation of uncertainty
\begin{frame}{Extreme example: TASEP}



\end{frame}


\begin{frame}{Propagation of dependency}

\end{frame}


% e^tG; previous work
\begin{frame}{Idealized solution}

  $e^{tG}$

\end{frame}

\begin{frame}{Previous work}

  list these

\end{frame}

% Assumptions and not
\begin{frame}{Assumptions}

  Assume:

    homogeneity


  Do not assume:

    stationarity

    reversibility

\end{frame}


%%%%%%%%% METHODS
\section{Statistical \& Computational Methods}

%   restrict to pattern frequencies: partial likelihood
\begin{frame}{Larger windows?}

  Pattern counting

  Partial/pseudo-likelihood.

\end{frame}


%   still missing edge effects: marginalize over window
%     node average boundary
\begin{frame}{Still missing those edge effects}

  Trim the bottom window.

\end{frame}


%   asymptotically correct: propagation of dependency
\begin{frame}{Asymptotic correctness}
  picture proof
\end{frame}

%   likelihood function
\begin{frame}{Inference framework: likelihood function}

  single branch; time is confounded with mutation

  given parameters:
    
    mutation patterns

    selection patterns (and form)


  and:

    set of patterns

    set of positions


  likelihood is: (multinomial)

\end{frame}

%   computation: sparse matrices
\begin{frame}{Computation: using sparseness}

\end{frame}

%   phylogenetic peeling (?)
\begin{frame}{Larger phylogenies}

\end{frame}


%%%%%%%%% APPLICATIONS
\section{Applications}

%   CpG
\begin{frame}{Nucleotide sequence}

\end{frame}

%   TASEP
\begin{frame}{Propagation of dependency: TASEP}

\end{frame}

%   Ising
\begin{frame}{Statistical physics: the Ising model}

\end{frame}

%   real data?
%   ancestral state reconstruction


%%%%% ok all done

\begin{frame}{Thanks}

  Matt Dean
  
  Rasmus Nielsen

  Simon Tavar\'e

  Graham Coop

  Sergey Nudzhin

\end{frame}

\end{document}
